internal/controller/controllers/image_controller.go

```diff
@@ -17,29 +17,154 @@ limitations under the License.
+
+	"github.com/go-openapi/strfmt"
+	"github.com/openshift/assisted-service/internal/bminventory"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-
-	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
 )
 
 // ImageReconciler reconciles a Image object
 type ImageReconciler struct {
 	client.Client
-	Log    logrus.FieldLogger
-	Scheme *runtime.Scheme
+	Log       logrus.FieldLogger
+	Scheme    *runtime.Scheme
+	Installer bminventory.InstallerInternals
 }
 
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=images,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=images/status,verbs=get;update;patch
 
 func (r *ImageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// your logic here
+	ctx := context.Background()
+
+	image := &adiiov1alpha1.Image{}
+	if err := r.Get(ctx, req.NamespacedName, image); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrap(err, "failed to get image data")
+	}
+
+	return r.create(ctx, image)
+}
+
+func (r *ImageReconciler) create(ctx context.Context, image *adiiov1alpha1.Image) (ctrl.Result, error) {
+	state := adiiov1alpha1.ImageStateFailedToCreate
+
+	staticIpConfig, staticIpErr := r.getStaticIpConfig(image)
+	if staticIpErr != nil {
+		return r.updateStatusAndReturnResult(ctx, image, nil, state, staticIpErr)
+	}
+
+	cluster, clusterRefErr := r.getClusterByRef(ctx, image.Spec.ClusterRef)
+	if clusterRefErr != nil {
+		return r.updateStatusAndReturnResult(ctx, image, nil, state, clusterRefErr)
+	}
+
+	var imageInfo *models.ImageInfo
+	updatedCluster, inventoryErr := r.Installer.GenerateClusterISOInternal(
+		ctx,
+		installer.GenerateClusterISOParams{
+			ClusterID: strfmt.UUID(cluster.Status.ID),
+			ImageCreateParams: &models.ImageCreateParams{
+				SSHPublicKey:    image.Spec.SSHPublicKey,
+				StaticIpsConfig: staticIpConfig,
+			},
+		})
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		state = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, image, imageInfo, state, inventoryErr)
+}
+
+func (r *ImageReconciler) getStaticIpConfig(image *adiiov1alpha1.Image) ([]*models.StaticIPConfig, error) {
+	var staticIpConfig []*models.StaticIPConfig
+	if image.Spec.StaticIpConfiguration == "" {
+		return staticIpConfig, nil
+	} else if err := json.Unmarshal([]byte(image.Spec.StaticIpConfiguration), &staticIpConfig); err != nil {
+		return nil, common.NewApiError(
```

Comment:
I think that it's a bit strange to use api errors for k8s api, @avishayt what do you think?